### ISSUE TITLE
Automate Building Update Site for On Disk Projects

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#689]
One requirement for On Disk Projects (#646, #669, #670) is to have a local update site available that contains the required Domino libraries.  This can be built from macOS or a Domino server using these instructions (from [this link](https://github.com/prominic/Moonshine-IDE/issues/669#issuecomment-595491928) ):

>Next, you need to build an update site.  A user should only need to do this once per workstation, and we'll look to see if there is a better way to do this for Moonshine.
>1. Clone this project:  https://github.com/OpenNTF/generate-domino-update-site
>2. Open the project in Terminal
>3. Run `mvn install`
>4. Run `mvn org.openntf.p2:generate-domino-update-site:generateUpdateSite -Ddest="/Users/-snip-/workspace/OnDiskProject/UpdateSite`, updating the path to match `<notes-platform>` above.
>
>This command automatically detects the Notes installation.  We'll probably want to manually compute the path from the Notes path configured for #667 using the `-Dsrc` parameter, but I need to find where the update site is stored in the installation

The Domino/Notes path can also be specified with the `-Dsrc` option.  For example, on a Linux Domino server (I also specified the version number to use the latest snapshot for this test):

    mvn org.openntf.p2:generate-domino-update-site:3.2.0-SNAPSHOT:generateUpdateSite -Dsrc="/opt/lotus/notes/latest/linux/" -Ddest="/root/test/UpdateSite"

NOTE:  We cannot provide a download links for update sites, because this will be a problem with the HCL Notes/Domino license.  The user needs to generate the update site from a copy of Notes or Domino that they own.

We need to add a section in the Domino settings page to specify the path to the update site.  This will allow the user to use a copy they have created elsewhere (for example, on a Domino server).  

This update site will be used for any On Disk Project Maven scripts.  Currently, this is configured in the global settings.xml file (see [this comment](https://github.com/prominic/Moonshine-IDE/issues/669#issuecomment-595491928) for more details).

We need to update this versionRange value in the base project pom.xml to match the update site.  If we can't detect this manually, this will also need to be a configured value:

```
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>target-platform-configuration</artifactId>
        <configuration>

          <pomDependencies>consider</pomDependencies>
          <dependency-resolution>
            <extraRequirements>
              <requirement>
                <type>eclipse-plugin</type>
                <id>com.ibm.notes.java.api.win32.linux</id>
                <versionRange>[11.0,11.1)</versionRange>
```

In addition, I would like to have a button or tool that the user can use to automatically generate the update site using the commands above.  This will require Git configuration.

**Comment by JoelProminic** [comment]
Unfortunately, I have not been able to successfully build an update site on Windows 10.  I have tested with IBM Domino 9.0.1 and 11.0.0.  We are still investigating this, and I'll update the issue if we find a solution.  See #646 for more details (various comments).

**Comment by JoelProminic** [comment]
We can start with the implementation for this for macOS.  I'll writeup some instructions to generate an update site from a Domino server manually, and we can display these so that Windows users can still generate the update site.

Add a new field called "Update Site" in the Domino settings, just under "HCL Notes Installation".  In addition to the Clear and Change buttons (still needed for manual or pre-existing updat sites), there should be a "Generate" button that is enabled on macOS only.  The Generate button will open a dialog like this, with buttons for "Generate" and "Cancel":

> Title:  Generate Update Site
>
> Generate a new update site based on your local HCL Notes installation.  This may take a couple minutes

Here are the commands to run.  Note that this will require Git and Maven to be configured in Moonshine, and we should use the full paths and environment variables as normal:

    git clone https://github.com/OpenNTF/generate-domino-update-site
    cd generate-domino-update-site
    mvn org.openntf.p2:generate-domino-update-site:generateUpdateSite -Dsrc="$NOTES_INSTALLATION_PATH/Contents/MacOS" -Ddest="$UPDATE_SITE_PATH"

NOTES_INSTALLATION_PATH is the "HCL Notes Installation" path collected in #667 (the full -Dsrc path should look something like this: `/Applications/HCL Notes.app/Contents/MacOS`).

UPDATE_SITE_PATH will be a new directory in a shared location.  I suggest putting this at `~/Downloads/MoonshineSDKs/Domino/UpdateSite`, but we could also potentially put it in the default Moonshine directory instead..  This path should be saved to te new "Update Site" settings field above so that it can be used to generate setting.xml later.  (Originally, we were using the global settings.xml, but I want to look into using a special settings.xml for Moonshine.  I'll give more details on this later).

It takes me about 56 seconds to generate the update site, but this may be faster or shorter for some users.  We should show at least a busy icon and script output on the console while this is running.

If the selected Update Site path already exists and is populated, report an error to the user to manually cleanup the old installation.  I'd like to avoid automatically deleting this for now.

> You have already generated an update site to $UPDATE_SITE_PATH.  Please delete or rename this directory and try again.

My TODO:
- Figure out how to handle update site in custom settings.xml
- Writeup manual instructions for Windows.

**Comment by rat-moonshine** [comment]
![image](https://user-images.githubusercontent.com/24370870/81261823-5d993700-905a-11ea-88da-feca6d54ec0e.png)

**Comment by JoelProminic** [comment]
I would like to have Moonshine generate a settings.xml file in a shared location (Moonshine internal files or ~/Documents/MoonshineSDKs/Domino).  This should be populated with the Notes Installation and Update Site settings:

    <?xml version="1.0"?>
    <settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0 http://maven.apache.org/xsd/settings-1.0.0.xsd">
        <profiles>
            <profile>
                <id>nsfodp</id>
                <properties>
                    
                    <notes-program>$NOTES_INSTALLATION_PATH/Contents/MacOS</notes-program>
                    <notes-platform>file://$UPDATE_SITE_PATH</notes-platform>
                </properties>
            </profile>
        </profiles>
        <activeProfiles>
            <activeProfile>nsfodp</activeProfile>
        </activeProfiles>
    </settings>

Note that the properties will be formatted differently for Windows.  TODO: example

This settings.xml file should be updated when the Domino settings are changed.

Then, we need to use this settings file with the Maven calls for the On Disk Project:

    mvn -s <settings_file> <commands>
    # Example:
    mvn -s ~/Downloads/MoonshineSDKs/Domino/settings.xml clean install

We already have a way to define a custom settings file in the On Disk Project Maven settings.  This should be populated with the default settings.xml path when creating a new Domino On Disk Project.  Since the file will be updated automatically, the user should not need to pay attention to this file unless they need to add their own settings.xml (they would need to remember to add the same properties to their own settings.xml in this case, but for now we'll assume that the user will know to do this if they are editing settings.xml):
![image](https://user-images.githubusercontent.com/24532731/81341525-4e74b200-9077-11ea-94dd-0fac56b82d7e.png)

I see the "Notes Programe Path" and "Notes Platform Path" settings here as well.  I don't remember why we added these properties originally, but I think they are obsolete with the updates from this issue.  We could potentially allow the use to define a different Notes Installation Path or Update Site Path for a project, but I don't think this is needed for now.

**Comment by JoelProminic** [comment]
Here is a template for settings.xml on Windows:

    <?xml version="1.0"?>
    <settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0 http://maven.apache.org/xsd/settings-1.0.0.xsd">
        <profiles>
            
            
            <profile>
                <id>nsfodp</id>
                <properties>
                    
                    <notes-program>$NOTES_INSTALLATION_PATH</notes-program>
                    <notes-platform>file:///$UPDATE_SITE_PATH</notes-platform>
                </properties>
            </profile>
            
            
        </profiles>
        <activeProfiles>
            <activeProfile>nsfodp</activeProfile>
        </activeProfiles>
    </settings>

Example values:
- NOTES_INSTALLATION_PATH:  C:/Program Files (x86)/IBM/Notes
- UPDATE_SITE_PATH:  C:/temp/OnDiskProject_0409/Domino10_Server_UpdateSite/UpdateSite

Compared to the MacOS version, "/Contents/MacOS/" is missing from notes-program and there is an extra '/' in notes-platform.

**Comment by rat-moonshine** [comment]
"Update Path" value is not updating immediately after a generation - fixed.

**Comment by rat-moonshine** [comment]
New settings file now generates based on Domino settings update-site save. Note, an existing setting file do not updates if the path set to 'Update Site' field is invalid or user cleared the path. Let me know if this make sense.

macOS path: ~/Downloads/MoonshineSDKs/Domino/settings.xml
Windows path: C:/MoonshineSDKs/Domino/settings.xml
(I couldn't able to give a test on Windows yet as my system broken)

New OnDisk project now auto-set it's 'Maven Settings File' path value with above location, if exists. 

Let me know if I missed out something.

Please, check.

**Comment by rat-moonshine** [comment]
I have removed the "Notes Programe Path" and "Notes Platform Path" settings options for OnDisk project from interface.

**Comment by piotrzarzycki21** [comment]
We need to test App Store version of Moonshine  for that issue, than fix potential issue and merge to master.

**Comment by piotrzarzycki21** [comment]
- Add Icon for On Disk Project - @JoelProminic

**Comment by rat-moonshine** [comment]
While testing recently on App Store version of the Moonshine, automatic generation of Update Site file fails with following error:

`xcrun: can not run in sandbox `

It appears that we had this trouble in past too, while running Git SDK executable in Moonshine. We able to overcome the problem by using Xcode version of Git or CommandLineTools. The corresponding command to generate Update Site has involvement of git-clone, so updating the Git executable path could be a solution against this. 

Upon updating Git path I able to pass the "xcrun" error in sandbox. Git clone works but process broke during Maven command execution. Here's the detail to command and output:

// command
```
mkdir '/Users/devsena/Library/Containers/com.moonshine-ide/Data/Library/Application Support/com.moonshine-ide/Local Store/dominoUpdateSiteGeneration' && cd '/Users/devsena/Library/Containers/com.moonshine-ide/Data/Library/Application Support/com.moonshine-ide/Local Store/dominoUpdateSiteGeneration' && /Applications/Xcode.app/Contents/Developer/usr/bin/git clone --progress -v https://github.com/OpenNTF/generate-domino-update-site . && cd generate-domino-update-site && /Users/devsena/Downloads/MoonshineSDKs/Maven/apache-maven-3.6.0/bin/mvn org.openntf.p2:generate-domino-update-site:generateUpdateSite -Dsrc="/Applications/HCL Notes.app/Contents/MacOS" -Ddest="/Users/devsena/Downloads/MoonshineSDKs/Domino/UpdateSite"
```

// output
```
: Cloning into '.'...
: POST git-upload-pack (583 bytes)
: remote: Enumerating objects: 585, done.        
: remote: Counting objects:   0% (1/585)        
: remote: Counting objects:  72% (422/585)        
: remote: Compressing objects:   0% (1/195)        
: remote: Compressing objects:  93% (182/195)        
: Receiving objects:   0% (1/585)   
: Receiving objects:  21% (123/585)   
: Receiving objects:  72% (422/585)   
: Receiving objects:  77% (451/585)   
: Receiving objects:  84% (492/585)   
: Receiving objects:  86% (504/585)   
: remote: Total 585 (delta 205), reused 553 (delta 173), pack-reused 0        
: Receiving objects:  90% (527/585)   
: Resolving deltas:   0% (0/205)   
: Resolving deltas:  53% (110/205)   
: [INFO] Scanning for projects...
: [INFO] ------------------------------------------------------------------------
: [INFO] BUILD FAILURE
: [INFO] ------------------------------------------------------------------------
: [INFO] Total time:  0.200 s
: [INFO] Finished at: 2020-05-14T14:11:23+05:30
: [INFO] ------------------------------------------------------------------------
: [ERROR] Error resolving version for plugin 'org.openntf.p2:generate-domino-update-site' from the repositories [local (/Users/devsena/Library/Con
: tainers/com.moonshine-ide/Data/.m2/repository), central (https://repo.maven.apache.org/maven2)]: Plugin not found in any plugin repository -> [Help 1]
: [ERROR] 
: [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
: [ERROR] Re-run Maven using the -X switch to enable full debug logging.
: [ERROR] 
: [ERROR] For more informati
: on about the errors and possible solutions, please read the following articles:
: [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginVersionResolutionException
```

@JoelProminic do you have any thought why there could be a `Plugin not found in any plugin repository` error (?)

**Comment by JoelProminic** [comment]
I forgot the `mvn install` command in my [writeup](https://github.com/prominic/Moonshine-IDE/issues/689#issuecomment-624934185).  

    git clone https://github.com/OpenNTF/generate-domino-update-site
    cd generate-domino-update-site
    mvn install
    mvn org.openntf.p2:generate-domino-update-site:generateUpdateSite -Dsrc="$NOTES_INSTALLATION_PATH/Contents/MacOS" -Ddest="$UPDATE_SITE_PATH"

We could probably instead get the dependency from the openntf Maven repository, but I need to investigate this more.  The above commands should work.

**Comment by JoelProminic** [comment]
@rat-moonshine reported that the `mvn install` command has been added.   I'll retest this.

**Comment by piotrzarzycki21** [comment]
Moving this for next release. No announcement in CHANGELOG

**Comment by rat-moonshine** [comment]
This also included in 'master'. However, due to its test complexity, we may move this to v3.1.0 and CHANGELOG doesn't require any explicit reference for this in v3.0.

**Comment by piotrzarzycki21** [comment]
If it's merged - close it.

**Comment by JoelProminic** [comment]
This was unofficially included in the 3.0.0 release.  I need to revisit it before the official release to make sure it is still working for macOS and Windows, and if there are any remaining issues or limitation I have forgotten.   I think the On Disk Project is not working properly for Windows in general, though.

**Comment by JoelProminic** [comment]
I did a test with generating the update site from a Vagrant VM with Domino installed.  This seemed to work well, so we could consider using Vagrant to generate the update site when we get the Vagrant integration working (#770):

    [vagrant@localhost generate-domino-update-site]$ mkdir test-update-site
    [vagrant@localhost generate-domino-update-site]$ cd test-update-site

    # Maven Install:  https://maven.apache.org/download.cgi
    [vagrant@localhost test-update-site]$ wget https://dlcdn.apache.org/maven/maven-3/3.8.4/binaries/apache-maven-3.8.4-bin.tar.gz
    ...
    [vagrant@localhost test-update-site]$ sha512sum apache-maven-3.8.4-bin.tar.gz 
    a9b2d825eacf2e771ed5d6b0e01398589ac1bfa4171f36154d1b5787879605507802f699da6f7cfc80732a5282fd31b28e4cd6052338cbef0fa1358b48a5e3c8  apache-maven-3.8.4-bin.tar.gz


    # Generate the update site
    [vagrant@localhost test-update-site]$ export MAVEN_HOME=/home/vagrant/test-update-site/apache-maven-3.8.4
    [vagrant@localhost test-update-site]$ export PATH=$MAVEN_HOME/bin:$PATH
    [vagrant@localhost generate-domino-update-site]$ mvn -s settings.xml org.openntf.p2:generate-domino-update-site:4.0.0:generateUpdateSite -Dsrc="/opt/ibm/domino/notes/latest/linux/" -Ddest="/home/vagrant/test-update-site/UpdateSite"
    ...


    # Get a zip of the update site
    [vagrant@localhost generate-domino-update-site]$ cd ~/test-update-site
    [vagrant@localhost test-update-site]$ zip -r /vagrant/UpdateSite.zip UpdateSite/ 

We already have a simpler method for macOS, but this could be useful for Windows, or if we move the On Disk Project work entirely to Vagrant.
